agent/checks/check.go

```diff
@@ -715,3 +703,50 @@ func (c *CheckGRPC) Stop() {
 		close(c.stopCh)
 	}
 }
+
+// StatusHandler keep tracks of successive error/success counts and ensures
+// that status can be set to critical/passing only once the successive number of event
+// reaches the given threshold.
+type StatusHandler struct {
+	inner                  CheckNotifier
+	logger                 *log.Logger
+	successBeforePassing   int
+	successCounter         int
+	failuresBeforeCritical int
+	failuresCounter        int
+}
+
+// NewStatusHandler set counters values to threshold in order to immediatly update status after first check.
+func NewStatusHandler(inner CheckNotifier, logger *log.Logger, successBeforePassing, failuresBeforeCritical int) *StatusHandler {
+	return &StatusHandler{
+		logger:                 logger,
+		inner:                  inner,
+		successBeforePassing:   successBeforePassing,
+		successCounter:         successBeforePassing,
+		failuresBeforeCritical: failuresBeforeCritical,
+		failuresCounter:        failuresBeforeCritical,
+	}
+}
+
+func (s *StatusHandler) updateCheck(checkID types.CheckID, status, output string) {
+
+	if status == api.HealthPassing || status == api.HealthWarning {
+		s.successCounter++
+		if s.successCounter >= s.successBeforePassing {
+			s.logger.Printf("[DEBUG] agent: Check %q is %q", checkID, status)
+			s.inner.UpdateCheck(checkID, status, output)
+			s.failuresCounter = 0
+			return
+		}
+		s.logger.Printf("[WARN] agent: Check %q was %q but has not reached success threshold %d/%d", checkID, status, s.successCounter, s.successBeforePassing)
+	} else {
+		s.failuresCounter++
+		if s.failuresCounter >= s.failuresBeforeCritical {
+			s.logger.Printf("[WARN] agent: Check %q is now critical", checkID)
+			s.inner.UpdateCheck(checkID, status, output)
+			s.successCounter = 0
+			return
+		}
+		s.logger.Printf("[WARN] agent: Check %q failed but has not reached failure threshold %d/%d", checkID, s.failuresCounter, s.failuresBeforeCritical)
+	}
```

Comment:
We never reset the counters which means that the thresholds will only work the first time round doesn't it?

e.g.  if you have `successBeforePassing = 2, failuresBeforeCritical = 2` then you get the sequence: OK, OK, OK, FAIL, FAIL, FAIL, OK

The check should not yet be passing again since there is only one OK since the last FAIL but the state here would be: `successCounter = 4, successBeforePassing = 2` and so `s.successCounter >= s.successBeforePassing` would be true.

Unless I'm missing something? I don't think there is a test case that exercised the path beyond a single threshold traversal which would presumably catch this issue.

I think it's really easy to fix though - just need to reset success counter when failure threshold hits and vice versa but some through table tests would be good here because if you reset it too soon you also break the thresholding effect. I think there are at least 6-8 interesting check histories to enumerate here and test that this does the right think in each to be confident in the logic so a table test seems like an efficient way to do that.

Reply:
Indeed the counter are not reset after different checks as they should ; counters are now set to 0 after updateCheck() with the opposite status is called, plus a test to ensure that sequences such as FAIL PASS FAIL FAIL won't update status to critical if threshold is set to 3
I think the current tests bring good coverage, but if you think there is added value to go for the full table we can go for it, your call.

Thanks for the catch !

Reply:
I agree those tests seem to cover the cases. I have a preference for table tests over multiple long story-based tests for cases like this because it's easier to read the table and see which cases are tested than having to keep state in your head as you read but I think this is fine for now.